# Assignment 10
## Association Rules

## Objective:

 The Objective of this assignment is to perform necessary operations like preprocessing, and Applying the Apriori algorithm to determine the association rules for the given dataset ("Online retail")

 

## Task 1 : Data Preprocessing 

In [1]:
# Import necessary libraries
import numpy as np 
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings("ignore")

In [2]:
# Load the dataset
data = pd.read_excel("online retail.xlsx")

In [3]:
data

shrimp,almonds,avocado,vegetables mix,green grapes,whole weat flour,yams,cottage cheese,energy drink,tomato juice,low fat yogurt,green tea,honey,salad,mineral water,salmon,antioxydant juice,frozen smoothie,spinach,olive oil
0                                burgers,meatballs,eggs                                                                                                                                                                             
1                                               chutney                                                                                                                                                                             
2                                        turkey,avocado                                                                                                                                                                             
3     mineral water,milk,energy bar,whole wheat rice...                                                                                                                                                                             
4                                        low fat yogurt                                                                                                                                                                             
...                                                 ...                                                                                                                                                                             
7495                      butter,light mayo,fresh bread                                                                                                                                                                             
7496  burgers,frozen vegetables,eggs,french fries,ma...                                                                                                                                                                             
7497                                            chicken                                                                                                                                                                             
7498                                 escalope,green tea                                                                                                                                                                             
7499    eggs,frozen smoothie,yogurt cake,low fat yogurt                                                                                                                                                                             

[7500 rows x 1 columns]

In [4]:
data.columns =["items"]

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7500 entries, 0 to 7499
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   items   7500 non-null   object
dtypes: object(1)
memory usage: 58.7+ KB


In [6]:
# So the data has no null values.
# we have to use transaction encoder for converting the transaction data into appropriate format.

# Converting the data into list of transactions.
Data = [ trans.split(',') for trans in data['items'] ]

In [7]:
Data

[['burgers', 'meatballs', 'eggs'],
 ['chutney'],
 ['turkey', 'avocado'],
 ['mineral water', 'milk', 'energy bar', 'whole wheat rice', 'green tea'],
 ['low fat yogurt'],
 ['whole wheat pasta', 'french fries'],
 ['soup', 'light cream', 'shallot'],
 ['frozen vegetables', 'spaghetti', 'green tea'],
 ['french fries'],
 ['eggs', 'pet food'],
 ['cookies'],
 ['turkey', 'burgers', 'mineral water', 'eggs', 'cooking oil'],
 ['spaghetti', 'champagne', 'cookies'],
 ['mineral water', 'salmon'],
 ['mineral water'],
 ['shrimp',
  'chocolate',
  'chicken',
  'honey',
  'oil',
  'cooking oil',
  'low fat yogurt'],
 ['turkey', 'eggs'],
 ['turkey',
  'fresh tuna',
  'tomatoes',
  'spaghetti',
  'mineral water',
  'black tea',
  'salmon',
  'eggs',
  'chicken',
  'extra dark chocolate'],
 ['meatballs', 'milk', 'honey', 'french fries', 'protein bar'],
 ['red wine', 'shrimp', 'pasta', 'pepper', 'eggs', 'chocolate', 'shampoo'],
 ['rice', 'sparkling water'],
 ['spaghetti', 'mineral water', 'ham', 'body spray',

In [8]:
# Applying Transaction Encoder
!pip install mlxtend

In [9]:
from mlxtend.preprocessing import TransactionEncoder
te = TransactionEncoder()

In [10]:
te_array = te.fit(Data).transform(Data)

In [11]:
te_array

array([[False, False, False, ..., False, False, False],
       [False, False, False, ..., False, False, False],
       [False, False, False, ..., False, False, False],
       ...,
       [False, False, False, ..., False, False, False],
       [False, False, False, ..., False, False, False],
       [False, False, False, ..., False,  True, False]])

In [12]:
df = pd.DataFrame(te_array,columns = te.columns_)

In [13]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7500 entries, 0 to 7499
Columns: 120 entries,  asparagus to zucchini
dtypes: bool(120)
memory usage: 879.0 KB


In [14]:
df.isnull().sum().sum()

0

## Task 2 : Association Rule Mining

In [15]:
# import apriori algorithm and find the frequent_itemsets

from mlxtend.frequent_patterns import apriori,association_rules

In [26]:
frequent_itemsets = apriori(df,min_support=0.03, use_colnames=True)

In [27]:
frequent_itemsets

support                            itemsets
0   0.033200                           (avocado)
1   0.033733                          (brownies)
2   0.087200                           (burgers)
3   0.030133                            (butter)
4   0.081067                              (cake)
5   0.046800                         (champagne)
6   0.060000                           (chicken)
7   0.163867                         (chocolate)
8   0.080400                           (cookies)
9   0.051067                       (cooking oil)
10  0.031733                    (cottage cheese)
11  0.179733                              (eggs)
12  0.079333                          (escalope)
13  0.170933                      (french fries)
14  0.043067                       (fresh bread)
15  0.063200                   (frozen smoothie)
16  0.095333                 (frozen vegetables)
17  0.052400                     (grated cheese)
18  0.132000                         (green tea)
19  0.098267                       (ground beef)
20  0.049467                     (herb & pepper)
21  0.047333                             (honey)
22  0.032400                          (hot dogs)
23  0.076400                    (low fat yogurt)
24  0.129600                              (milk)
25  0.238267                     (mineral water)
26  0.065733                         (olive oil)
27  0.095067                          (pancakes)
28  0.042400                            (salmon)
29  0.071333                            (shrimp)
30  0.050533                              (soup)
31  0.174133                         (spaghetti)
32  0.030267                      (tomato juice)
33  0.068400                          (tomatoes)
34  0.062533                            (turkey)
35  0.058533                  (whole wheat rice)
36  0.033200                   (chocolate, eggs)
37  0.034400           (chocolate, french fries)
38  0.032133                   (chocolate, milk)
39  0.052667          (chocolate, mineral water)
40  0.039200              (chocolate, spaghetti)
41  0.036400                (eggs, french fries)
42  0.030800                        (eggs, milk)
43  0.050933               (eggs, mineral water)
44  0.036533                   (eggs, spaghetti)
45  0.033733       (french fries, mineral water)
46  0.035733  (frozen vegetables, mineral water)
47  0.030933          (green tea, mineral water)
48  0.040933        (mineral water, ground beef)
49  0.039200            (ground beef, spaghetti)
50  0.048000               (mineral water, milk)
51  0.035467                   (milk, spaghetti)
52  0.033733           (mineral water, pancakes)
53  0.059733          (mineral water, spaghetti)

In [28]:
rules = association_rules(frequent_itemsets,metric="lift",min_threshold=0.5)

In [32]:
rules.sort_values(by = "confidence")

antecedents          consequents  antecedent support  \
23      (mineral water)          (green tea)            0.238267   
32      (mineral water)           (pancakes)            0.238267   
19      (mineral water)       (french fries)            0.238267   
21      (mineral water)  (frozen vegetables)            0.238267   
12               (eggs)               (milk)            0.179733   
24      (mineral water)        (ground beef)            0.238267   
1                (eggs)          (chocolate)            0.179733   
4           (chocolate)               (milk)            0.163867   
18       (french fries)      (mineral water)            0.170933   
3        (french fries)          (chocolate)            0.170933   
28      (mineral water)               (milk)            0.238267   
10               (eggs)       (french fries)            0.179733   
0           (chocolate)               (eggs)            0.163867   
16               (eggs)          (spaghetti)            0.179733   
31          (spaghetti)               (milk)            0.174133   
17          (spaghetti)               (eggs)            0.174133   
2           (chocolate)       (french fries)            0.163867   
11       (french fries)               (eggs)            0.170933   
15      (mineral water)               (eggs)            0.238267   
7       (mineral water)          (chocolate)            0.238267   
9           (spaghetti)          (chocolate)            0.174133   
27          (spaghetti)        (ground beef)            0.174133   
22          (green tea)      (mineral water)            0.132000   
13               (milk)               (eggs)            0.129600   
8           (chocolate)          (spaghetti)            0.163867   
5                (milk)          (chocolate)            0.129600   
34      (mineral water)          (spaghetti)            0.238267   
30               (milk)          (spaghetti)            0.129600   
14               (eggs)      (mineral water)            0.179733   
6           (chocolate)      (mineral water)            0.163867   
35          (spaghetti)      (mineral water)            0.174133   
33           (pancakes)      (mineral water)            0.095067   
29               (milk)      (mineral water)            0.129600   
20  (frozen vegetables)      (mineral water)            0.095333   
26        (ground beef)          (spaghetti)            0.098267   
25        (ground beef)      (mineral water)            0.098267   

    consequent support   support  confidence      lift  leverage  conviction  \
23            0.132000  0.030933    0.129827  0.983534 -0.000518    0.997502   
32            0.095067  0.033733    0.141578  1.489250  0.011082    1.054182   
19            0.170933  0.033733    0.141578  0.828265 -0.006994    0.965803   
21            0.095333  0.035733    0.149972  1.573133  0.013019    1.064279   
12            0.129600  0.030800    0.171365  1.322261  0.007507    1.050402   
24            0.098267  0.040933    0.171796  1.748266  0.017520    1.088782   
1             0.163867  0.033200    0.184718  1.127246  0.003748    1.025576   
4             0.129600  0.032133    0.196094  1.513074  0.010896    1.082714   
18            0.238267  0.033733    0.197348  0.828265 -0.006994    0.949021   
3             0.163867  0.034400    0.201248  1.228121  0.006390    1.046800   
28            0.129600  0.048000    0.201455  1.554436  0.017121    1.089982   
10            0.170933  0.036400    0.202522  1.184803  0.005678    1.039611   
0             0.179733  0.033200    0.202604  1.127246  0.003748    1.028681   
16            0.174133  0.036533    0.203264  1.167290  0.005236    1.036563   
31            0.129600  0.035467    0.203675  1.571569  0.012899    1.093022   
17            0.179733  0.036533    0.209801  1.167290  0.005236    1.038051   
2             0.170933  0.034400    0.209927  1.228121  0.006390    1.049354   
11            0.179733  0.036400    0.212949  1.184803 

## Task 3 : Analysis and Interpretation: 

(chocolate) → (eggs)

**Support**: 3.32% of transactions contain both chocolate and eggs.

**Confidence**: 20.26% of transactions that contain chocolate also contain eggs.

**Lift**: 1.13, meaning that chocolate increases the likelihood of eggs being purchased by about 13% compared to random chance.

(chocolate) → (french fries)

**Support**: 3.44% of transactions contain both chocolate and french fries.

**Confidence**: 20.99% of transactions that contain chocolate also contain french fries.

**Lift**: 1.23, indicating a stronger association between chocolate and french fries compared to the previous rule.

(milk) → (chocolate)

**Support**: 3.21% of transactions contain both milk and chocolate.

**Confidence**: 24.79% of transactions that contain milk also contain chocolate.

**Lift**: 1.51, indicating a much stronger association between milk and chocolate.


In [33]:
## Saving the Association Rules 
rules.to_csv("rules.csv")

antecedents          consequents  antecedent support  \
0           (chocolate)               (eggs)            0.163867   
1                (eggs)          (chocolate)            0.179733   
2           (chocolate)       (french fries)            0.163867   
3        (french fries)          (chocolate)            0.170933   
4           (chocolate)               (milk)            0.163867   
5                (milk)          (chocolate)            0.129600   
6           (chocolate)      (mineral water)            0.163867   
7       (mineral water)          (chocolate)            0.238267   
8           (chocolate)          (spaghetti)            0.163867   
9           (spaghetti)          (chocolate)            0.174133   
10               (eggs)       (french fries)            0.179733   
11       (french fries)               (eggs)            0.170933   
12               (eggs)               (milk)            0.179733   
13               (milk)               (eggs)            0.129600   
14               (eggs)      (mineral water)            0.179733   
15      (mineral water)               (eggs)            0.238267   
16               (eggs)          (spaghetti)            0.179733   
17          (spaghetti)               (eggs)            0.174133   
18       (french fries)      (mineral water)            0.170933   
19      (mineral water)       (french fries)            0.238267   
20  (frozen vegetables)      (mineral water)            0.095333   
21      (mineral water)  (frozen vegetables)            0.238267   
22          (green tea)      (mineral water)            0.132000   
23      (mineral water)          (green tea)            0.238267   
24      (mineral water)        (ground beef)            0.238267   
25        (ground beef)      (mineral water)            0.098267   
26        (ground beef)          (spaghetti)            0.098267   
27          (spaghetti)        (ground beef)            0.174133   
28      (mineral water)               (milk)            0.238267   
29               (milk)      (mineral water)            0.129600   
30               (milk)          (spaghetti)            0.129600   
31          (spaghetti)               (milk)            0.174133   
32      (mineral water)           (pancakes)            0.238267   
33           (pancakes)      (mineral water)            0.095067   
34      (mineral water)          (spaghetti)            0.238267   
35          (spaghetti)      (mineral water)            0.174133   

    consequent support   support  confidence      lift  leverage  conviction  \
0             0.179733  0.033200    0.202604  1.127246  0.003748    1.028681   
1             0.163867  0.033200    0.184718  1.127246  0.003748    1.025576   
2             0.170933  0.034400    0.209927  1.228121  0.006390    1.049354   
3             0.163867  0.034400    0.201248  1.228121  0.006390    1.046800   
4             0.129600  0.032133    0.196094  1.513074  0.010896    1.082714   
5             0.163867  0.032133    0.247942  1.513074  0.010896    1.111794   
6             0.238267  0.052667    0.321400  1.348907  0.013623    1.122506   
7             0.163867  0.052667    0.221041  1.348907  0.013623    1.073398   
8             0.174133  0.039200    0.239219  1.373768  0.010665    1.085551   
9             0.163867  0.039200    0.225115  1.373768  0.010665    1.079042   
10            0.170933  0.036400    0.202522  1.184803  0.005678    1.039611   
11            0.179733  0.036400    0.212949  1.184803  0.005678    1.042202   
12            0.129600  0.030800    0.171365  1.322261  0.007507    1.050402   
13            0.179733  0.030800    0.237654  1.322261  0.007507    1.075977   
14            0.238267  0.050933    0.283383  1.189351  0.008109    1.062957   
15            0.179733  0.050933    0.213766  1.189351  0.008109    1.043286   
16            0.174133  0.036533    0.203264  1.167290  0.005236    1.036563   
17            0.179733  0.036533    0.209801  1.167290 

## Interview questions : 
### 1.	What is lift and why is it important in Association rules?

   **Lift** : lift is the ratio of the observed support to that expected if X and Y are independent .
   it measures how much more likely the consequent is given the antcecedent compared to its typical occurances.

### 2. what is support and Confidence, how do you calculate them ?

   **Support** : measures how frequently the itemset appears in the dataset.

   **support** (X) = no.of transactions contain X/ total no.of transactions.

   **Confidence** : the probability that a transaction containing antecedent also contains the consequent.

   **confidence**(x-->y) = support(XUY)/support(x)

### 3. What are some limitations or challenges of Association mining ?

1. Too Many Rules : It generates large number of rules, many of them are irrelevant and uninteresting.
2. Support for Rare Items : items with low frequency may be overlooked, even if they have strong associations.
3. Difficulty in Setting Threshold values.
